pkl-executor/src/main/java/org/pkl/executor/ExecutorOptions.java

```diff
@@ -15,47 +15,226 @@
  */
 package org.pkl.executor;
 
+import java.net.URI;
 import java.nio.file.Path;
 import java.time.Duration;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import org.pkl.executor.spi.v1.ExecutorSpiOptions;
+import org.pkl.executor.spi.v1.ExecutorSpiOptions2;
 
 /**
  * Options for {@link Executor#evaluatePath}.
  *
- * <p>Note that subclasses of {@code ExecutorOptions} offer additional options.
+ * <p>To create {@code ExecutorOptions}, use its {@linkplain #builder builder}.
  */
-public class ExecutorOptions {
-  protected final List<String> allowedModules;
+public final class ExecutorOptions {
+  private final List<String> allowedModules;
 
-  protected final List<String> allowedResources;
+  private final List<String> allowedResources;
 
-  protected final Map<String, String> environmentVariables;
+  private final Map<String, String> environmentVariables;
 
-  protected final Map<String, String> externalProperties;
+  private final Map<String, String> externalProperties;
 
-  protected final List<Path> modulePath;
+  private final List<Path> modulePath;
 
-  protected final /* @Nullable */ Path rootDir;
+  private final /* @Nullable */ Path rootDir;
 
-  protected final /* @Nullable */ Duration timeout;
+  private final /* @Nullable */ Duration timeout;
 
-  protected final /* @Nullable */ String outputFormat;
+  private final /* @Nullable */ String outputFormat;
 
-  protected final /* @Nullable */ Path moduleCacheDir;
+  private final /* @Nullable */ Path moduleCacheDir;
 
-  protected final /* @Nullable */ Path projectDir;
+  private final /* @Nullable */ Path projectDir;
+
+  private final List<Path> certificateFiles;
+
+  private final List<URI> certificateUris;
+
+  private final int testPort; // -1 means disabled
+
+  private final int spiOptionsVersion; // -1 means use latest
 
   /** Returns the module cache dir that the CLI uses by default. */
   public static Path defaultModuleCacheDir() {
     return Path.of(System.getProperty("user.home"), ".pkl", "cache");
   }
 
+  public static Builder builder() {
+    return new Builder();
+  }
+
```

Comment:
Good idea with the builder!

We have an existing pattern of having `unconfigured()` and `preconfigured()` builders; see `EvaluatorBuilder`, `ConfigEvaluatorBuilder`, `ValueMapperBuilder`, etc. I think let's follow that.

```suggestion
  // keep in sync with SecurityManagers.defaultAllowedModules
  /// Returns the allowed module patterns that the CLI uses by default.
  public static final List<String> defaultAllowedModules =
    List.of(
      "repl:",
      "file:",
      "jar:file:",
      "modulepath:",
      "https:",
      "pkl:",
      "package:",
      "projectpackage:"
    );

  // keep in sync with SecurityManagers.defaultAllowedResources
  /// Returns the allowed resources patterns that the CLI uses by default.
  public static final List<String> defaultAllowedResources =
    List.of(
      "prop:",
      "env:",
      "file:",
      "modulepath:",
      "package:",
      "projectpackage:",
      "https:");

  /**
   * Creates a builder that has no options set.
   */
  public static Builder unconfigured() {
    return new Builder();
  }

  /**
   * Creates a builder configured with:
   *
   * <ul>
   *   <li>{@link #defaultAllowedModules}</li>
   *   <li>{@link #defaultAllowedResources}</li>
   *   <li>System environment variables</li>
   *   <li>System properties</li>
   * </ul>
   */
  public static Builder preconfigured() {
    //noinspection unchecked,rawtypes
    return unconfigured()
      .allowedModules(defaultAllowedModules)
      .allowedResources(defaultAllowedResources)
      .environmentVariables(System.getenv())
      .externalProperties((Map) System.getProperties());
  }
```